### PR TITLE
Move repeat autograd to C++.

### PR DESCRIPTION
**Pull request by gchanan**
This is necessary to support scalars.

### CODE REVIEW

File: tools/autograd/templates/Functions.cpp [hunk 1]
```diff
@@ -401,6 +401,33 @@ Tensor renorm_backward(const Tensor & grad, const Tensor & self, Scalar p, int64
...
+  for (int64_t i = 0; i < num_unsqueezed; ++i) {
+    grad = grad.sum(0, false);
+  }
+  for (size_t j = num_unsqueezed; j < repeats.size(); ++j) {
+    int64_t repeat = repeats[j];
+    if (repeat == 1) {
+      continue;
+    }
+    int64_t dim = j - num_unsqueezed;
+    grad = sum_tensorlist(grad.chunk(repeat, dim));
```
Comment (apaszke) [inline tools/autograd/templates/Functions.cpp]: Actually couldn't we implement this as a single `sum(dim=...)` call? I think that if you have a 4D tensor with sizes `(s1, s2, s3, s4)` and repeat it like this: `(2, 3, 4, 5)`, then you effectively could take a 4D result you got, view it like this: `(s1, 2, s2, 3, s3, 4, s4, 5)` and then call `.sum(dim=i)` for even `i`.

Comment (gchanan) [reply tools/autograd/templates/Functions.cpp]: that should work, we can even fold the first for loop in as well.

**Review by colesbury — APPROVED** [review]
Looks reasonable. I don't really understand the old repeat logic, but it looks like this matches it.

File: tools/autograd/templates/Functions.cpp (hunk 1 above)
Comment (gchanan) [reply tools/autograd/templates/Functions.cpp]: hmm, I thought we had multi-dimensional `sum` but I guess we don't yet (only squeeze?).  I'll defer this until then.

**Pull request merged**